### DISCUSSION TITLE
RDS2 brainstorming

### DISCUSSION THREAD

**Discussion by windytan** [windytan/redsea#102] [Ideas]
RDS2 support has been requested. Indeed it would be fun to have at some point, albeit quite a bit of work, depending on the approach. Ideas for how to realize the best possible UX are welcome in this thread!
## Complications
* RDS2 would be a diversion from how redsea currently works.
* Currently, redsea outputs numbers or very short strings in JSON line objects. These lines are normally < 150 characters long. Every JSON line is self-contained.
* RDS2 can send files of 163 kB in size, according to [radioworld](https://www.radioworld.com/tech-and-gear/whats-new-with-rds2).
### 1. Print Base64 to terminal?
163 kB in base64 would be ~217,000 characters on a single line, over a thousand times longer than the average redsea output line. If we print it in terminal it's 52 times the size of a 120x35 window; arguably not very good UX.
### 2. Write it on disk?
Another way would be to write the files on disk, separate from the main JSON output. Currently there's no file output in redsea, so this would be the most involved approach. Some things would need to be thought out first:
* Does RDS2 or the user give us file names?
* Should a newly received file always rewrite the old one?
* Or should there be a numbering scheme?
* Currently redsea does not exit on errors, except if the startup fails. What if the write starts failing / disk becomes full?
* Should redsea avoid filling up the disk after a certain percentage? Do we need a cross-platform way to request disk usage metrics?
* What if two channels (`-c`) wish to write into a file with the same name?
* How can we reliably unit test this?
### 3. Limited RDS2 support?
Yet another way to realize RDS2 would be just to detect the additional subcarriers and perhaps print some metadata, without actually decoding the files.

**Comment by andimik** [comment]
In a first step, I prefer idea 3.

Reason: Some enthusiasts are using `redsea` in a Raspberry PI recording FM (and DAB) either on a fixed place but also mobile (in combination with a GPS mouse). 

https://github.com/hayguen/fmlist_scan

It would be good to have **at least** the information (for example thanks to the JSON output) that there is an additional carrier for RDS2. In a second step the authors (some radio friends of mine) need to modify his code.

Like: 

```
{"pi":"0x1234","rds2":"stream1","ODA":"station_logo"}
```

(Station logo is 0xFF7F in RDS2 block B), see for example:

```
#S1 8000 FF7F 0904 1A74 @2024/02/19 22:33:02.91
```

see https://www.rds.org.uk/2010/pdf/R22_039_1_pw72.pdf (password is free of charge, but to be requested per mail from RDS forum, Mr. Kopitz)

In case you are providing a first try in a separate github branch, I can test it in my lab at home, as I can generate a RDS2 signal (MPX recording of existing Radio Gioconda from Italy) as well.

**Comment by sm3ulc** [comment]
Hi!

My suggestion would to use base64-blobs inside json for files and other
binary objects like logos etc. Just to keep it simple.

Kind regards,

// David

Den tis 20 feb. 2024 21:36Oona Räisänen ***@***.***> skrev:

**Reply by sm3ulc** [reply]
To keep a nice UX a switch for only metadata (name, size) or vice versa could b e used.

**Comment by andimik** [comment]
Well, I've changed `constexpr float kCarrierFrequency_Hz` to `66500.0f` (this is in file `subcarrier.cc`).

Now I can at least export the 1st stream:

```
$ cat /tmp/rds2-mpx-6255.wav | ./src/redsea -x -r 192k | head
---- FF7F 0904 1A74
---- 1100 0072 5110
2000 1200 0100 0000
---- FF7F 120A 9495
8000 FF7F 0904 1A74
---- 1704 0000 0001
2000 1800 0000 0000
---- FF7F 120D 43A1
8000 FF7F 0904 1A74
---- 1D61 696E 742E
```
[rds2-mpx-6255.zip](https://github.com/windytan/redsea/files/14411680/rds2-mpx-6255.zip)


or

```
$ more /tmp/53c5_rds2.wav | ./src/redsea -x -r 192k | head
---- ---- 1FB3 CA79
---- 33CA 577B ----
---- ---- 01A7 8108
---- 27E5 C3FE C01A
---- 71D5 DE11 A501
---- 6B43 EF08 538D
---- AEAE F101 5D08
---- 9498 343F 2EE1
---- FC4B 6868 FC4C
---- 8A14 9FBC F1BE
```
[53c5_rds2.zip](https://github.com/windytan/redsea/files/14411669/53c5_rds2.zip)


For RDS-only carriers there is no output.

**Comment by vaizki** [comment]
Idea for files: for each file, exec a configurable program, pipe the file data to its stdin and if the program prints out a valid JSON object to stdout then add that to the redsea output JSON. This way everyone can make their own file processing.

**Reply by sm3ulc** [reply]
Care to elaborate? RFT implemented in redsea? RFT in the mentioned exec?

**Comment by andimik** [comment]
I've added another recording from Australia to https://github.com/windytan/redsea/issues/32

**Comment by andimik** [comment]
@windytan

Oona:  any comments on our suggestions?

**Reply by windytan** [reply]
I think base64 is an interesting idea, raw hex output is also an interesting idea. Basically it's waiting for someone to pick up the task and write the implementation and tests :)

**Reply by windytan** [reply]
Or we could just detect the subcarriers and print metadata.

I would like to have this testing requirement for all new features implemented after 1.0, to catch obvious bugs and regressions; less work in the long run. The simplest test case would probably involve adding a very short clip of RDS2 MPX that we would run through redsea in Github Actions CI, and see that the JSON contains everything we want.

We haven't stored any test MPX in the repository before, but I'm thinking [Git-LFS](https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-git-large-file-storage) could be a relatively painless solution. I've used Git-LFS for [stereodemux tests](https://github.com/windytan/stereodemux/tree/master/test). The only downside is it seems to take a relatively long time to push new changes every time (is it uploading the files each time? I'm not sure) and also there is 1 gigabyte bandwidth limit per month for these hobby projects, including CI traffic and PRs, maybe even including forks. We should filter out any radio station audio to avoid copyright legal issues.

The build and static analysis [pipeline](https://github.com/windytan/redsea/blob/master/.github/workflows/build.yml) is set up so that it will automatically check all pull requests for the master branch, and show the green check mark if the PR builds on Ubuntu, old Debian, macOS, and Windows without issue. Currently runs in 7 minutes. If CI does not accept the pull request, clicking on the red X and then going to Details will reveal some useful logs.

There is a somewhat strict clang-tidy policy in place, but clang-tidy (and clang-format) can also be run locally before making PR.

Hope this helps possible contributors! I'll help with the PR for sure. I myself may have some time for new features later this year, unsure.

**Reply by windytan** [reply]
I also added the AC to the issue thread, because some people (me) like to work with such clear requirements: https://github.com/windytan/redsea/issues/32#issuecomment-2259896559

**Comment by windytan** [comment]
Hi, pull request #127 is up for testing for anyone interested. It can receive a file but will not check it for CRC just yet. I will make a few more revisions and rewrite the branch history a couple of times before merging. If you feel like reviewing the changes to find possible bugs etc. that's also of course welcome.

**Comment by andimik** [comment]
I've tested it with some RDS2 samples

# Breeze (Australia)

```
./redsea --streams -f rds2_mpx/Breeze\ 100.6\ MPX\ 17May2024.wav -r 192k | head
warning: ignoring sample rate parameter
{"stream":0,"pi":"0x49B1","group":"15A","tp":false,"prog_type":"Easy listening"}
{"stream":1,"group":"C","rft":{"pipe":0,"toggle":0,"byte_address":9050,"segment_data":[20,168,174,19,131]}}
{"stream":2,"group":"C","rft":{"pipe":0,"toggle":0,"byte_address":9045,"segment_data":[26,105,136,216,51]}}
{"stream":3,"group":"C","rft":{"pipe":0,"toggle":0,"byte_address":9055,"segment_data":[14,97,245,51,12]}}
{"stream":0,"pi":"0x49B1","group":"2A","tp":false,"prog_type":"Easy listening"}
{"stream":1,"group":"C","rft":{"pipe":0,"toggle":0,"byte_address":9065,"segment_data":[218,0,8,1,3]}}
{"stream":2,"group":"C","rft":{"pipe":0,"toggle":0,"byte_address":9060,"segment_data":[0,163,254,206,255]}}
{"stream":3,"group":"C","rft":{"pipe":0,"toggle":0,"byte_address":9070,"segment_data":[3,1,63,16,255]}}
{"stream":0,"pi":"0x49B1","group":"0A","tp":false,"prog_type":"Easy listening","di":{"stereo":true},"ta":false,"is_music":false}
{"stream":1,"group":"C","rft":{"pipe":0,"toggle":0,"byte_address":9080,"segment_data":[104,178,174,130,218]}}
```

![rft_file bin](https://github.com/user-attachments/assets/b393221c-449e-4371-ae97-4b1893ba01a5)

# Radio Gioconda (Italy)

For my Italian file it's not working, but I learned that they are using an outdated RDS2 specification (RPP data, see https://pira.cz/forum/index.php?PHPSESSID=dldv24aendfalqih0s8epiedcj&topic=936.0 )

```
$ ./redsea --streams -f /media/andreas/Elements/rds2_mpx/gioconda_20230915_udine_192khz_mono.wav -r 192k
warning: ignoring sample rate parameter
{"stream":0,"pi":"0x53C5","channel":1,"group":"2A","tp":true,"prog_type":"Pop music"}
{"stream":1,"channel":1,"group":"C","debug":["TODO: FID 2 FN 8"],"data":[140,247,51,241,105,128]}
{"stream":0,"pi":"0x53C5","channel":1,"group":"0A","tp":true,"prog_type":"Pop music","di":{"dynamic_pty":false},"ta":false,"is_music":true}
{"stream":1,"channel":1,"group":"C","debug":["TODO: FID 2 FN 8"],"data":[88,216,58,158,94,172]}
{"stream":0,"pi":"0x53C5","channel":1,"group":"3A","tp":true,"prog_type":"Pop music","open_data_app":{"oda_group":"12A","app_name":"RadioText+ (RT+)"}}
{"stream":1,"channel":1,"group":"C","debug":["TODO: FID 2 FN 8"],"data":[74,41,37,73,227,30]}
{"stream":0,"pi":"0x53C5","channel":1,"group":"0A","tp":true,"prog_type":"Pop music","di":{"compressed":false},"ta":false,"is_music":true}
{"stream":1,"channel":1,"group":"C","debug":["TODO: FID 2 FN 8"],"data":[207,178,231,234,152,122]}
{"stream":0,"pi":"0x53C5","channel":1,"group":"0A","tp":true,"prog_type":"Pop music","di":{"artificial_head":false},"ta":false,"is_music":true}
{"stream":1,"channel":1,"group":"C","debug":["TODO: FID 2 FN 8"],"data":[222,207,58,121,129,42]}
{"stream":0,"pi":"0x53C5","channel":1,"group":"0A","tp":true,"prog_type":"Pop music","di":{"stereo":true},"ta":false,"is_music":true,"alt_frequencies_b":{"tuned_frequency":106900,"same_programme":[104800]},"ps":"GIOCONDA"}
{"stream":1,"channel":1,"group":"C","debug":["TODO: FID 2 FN 8"],"data":[235,30,201,22,59,198]}
```

This would be the full content:

[gioconda_20230915_udine_192khz_mono_4min.txt](https://github.com/user-attachments/files/19168057/gioconda_20230915_udine_192khz_mono_4min.txt)


For me it's ok that this is not supported as this is not specified.

# Finland (Järviradio)

Unfortunately, my sample is not error-free.

[rds2-mpx-6255_test.wav.spy.txt](https://github.com/user-attachments/files/19168121/rds2-mpx-6255_test.wav.spy.txt)

# redsea sample

`rds2-minirds-192k.flac` works fine as well, see https://github.com/windytan/redsea/blob/dev/wip_streams/test/resources/rds2-minirds-192k.flac 
![rft_file bin](https://github.com/user-attachments/assets/832e1f12-9002-4056-8e3e-9e8e918cc228)

**Reply by windytan** [reply]
Thanks for testing!

**Reply by windytan** [reply]
It would be fun to support this RPP thing as well. With that gioconda file, seems if we just cut these bytes out by hand and put them into a file one after the other we get a JPEG picture. Then there's a 3-bit counter and a mystery 5-bit field.

                  ┌ ??        ┌  Picture data
                ┌─┴─┐    ┌────┴───────────────────────────────────────────┐
    30  1000100001110000 1111111111011000 1111111111100000 0000000000010000  8870 ffd8 ffe0 0010 .p......
    31  1000100010010001 0100101001000110 0100100101000110 0000000000000001  8891 4a46 4946 0001 ..JFIF..
    32  1000100010110010 0000000100000001 0000000100101100 0000000100101100  88b2 0101 012c 012c .....,.,
                     └┬┘
                      └ counter 0..7

![j2](https://github.com/user-attachments/assets/735c3b20-7fe0-4d29-8225-e06625a09fc8)

But I don't completely understand how to automatically find out if we missed any lines. Or where the file starts and ends. A fun puzzle for someone skilled in reverse engineering?

**Reply by andimik** [reply]
Oh, nice, you found the picture. 

Start and end? I thought there is a field showing the position.

**Comment by andimik** [comment]
Could you test with Järviradio?

```
$ cat ./6255_-_2025-03-12_23-04-56.spy | redsea -h --streams |  grep file_contents |  head --lines=1 |  jq --raw-output .rft.file_contents |  base64 --decode > rft_file.bin
```

The bin file which I got is 0 Bytes, but I expect the logo (which I could extract in RDSSpy).

I also tried the wav file (which is too big for Github).

```
$ cat ./6255_-_2025-03-12_23-04-56.spy | redsea -h --streams | head
{"stream":0,"pi":"0x6255","group":"14A","tp":false,"prog_type":"National music","other_network":{"pi":"0x6203","tp":true,"prog_type":"No PTY","ta":false}}
{"stream":1,"group":"C","rft":{"pipe":0,"toggle":0,"byte_address":1970,"segment_data":[28,56,217,252,36]}}
{"stream":0,"pi":"0x6255","group":"14A","tp":false,"prog_type":"National music","other_network":{"pi":"0x6202","tp":false,"prog_type":"Pop music","ta":true}}
{"stream":1,"group":"C","rft":{"pipe":0,"toggle":0,"byte_address":1975,"segment_data":[120,237,177,243,6]}}
{"stream":0,"pi":"0x6255","group":"15A","tp":false,"prog_type":"National music"}
{"stream":1,"group":"C","rft":{"pipe":0,"toggle":0,"byte_address":1980,"segment_data":[181,21,43,146,189]}}
{"stream":0,"pi":"0x6255","group":"12A","tp":false,"prog_type":"National music","unknown_oda":{"raw_data":"18 8022 0A9A"}}
{"stream":1,"group":"C","rft":{"pipe":0,"toggle":0,"byte_address":1985,"segment_data":[153,185,167,199,62]}}
{"stream":0,"pi":"0x6255","group":"1A","tp":false,"prog_type":"National music","prog_item_number":4369,"prog_item_started":{"day":2,"time":"04:17"},"has_linkage":false,"country":"fi"}
{"stream":1,"group":"C","rft":{"pipe":0,"toggle":0,"byte_address":1990,"segment_data":[136,86,217,63,51]}}
```

[6255_-_2025-03-12_23-04-56.spy.zip](https://github.com/user-attachments/files/19218769/6255_-_2025-03-12_23-04-56.spy.zip)
[6255_short_sorted.spy.zip](https://github.com/user-attachments/files/19218825/6255_short_sorted.spy.zip)

**Reply by windytan** [reply]
Nice, you found a bug :) It didn't work for files whose size is exactly divisible by 5. Fix coming a bit later.

**Comment by andimik** [comment]
By chance, I found another station on a Spyserver in Greece :-) and confirm, that all 3 streams work with `redsea`.

```
{"stream":0,"pi":"0x10C4","group":"2A","tp":true,"prog_type":"Pop music"}
{"stream":1,"group":"C","rft":{"pipe":0,"app_name":"RFT: Station logo","toggle":0,"byte_address":1970,"segment_data":[109,112,109,101,116]}}
{"stream":2,"group":"C","rft":{"pipe":0,"app_name":"RFT: Station logo","toggle":0,"byte_address":1965,"segment_data":[60,47,120,58,120]}}
{"stream":3,"group":"C","rft":{"pipe":0,"app_name":"RFT: Station logo","toggle":0,"byte_address":1975,"segment_data":[97,62,10,60,63]}}
{"stream":2,"group":"C","rft":{"pipe":0,"app_name":"RFT: Station logo","toggle":0,"byte_address":1995,"segment_data":[63,62,43,54,177]}}
{"stream":0,"pi":"0x10C4","group":"0A","tp":true,"prog_type":"Pop music","di":{"artificial_head":false},"ta":true,"is_music":true,"ps":"KARAMELA"}
```

![rds2_greece_10C4_karamela](https://github.com/user-attachments/assets/ee361c98-5ca8-47ad-969e-c640d62b855c)
[rds2_greece_10C4_karamela.spy.txt](https://github.com/user-attachments/files/19252433/rds2_greece_10C4_karamela.spy.txt)
(the original recording was quite painful due to massive errors, so I've cropped the attached file and suppressed missing blocks, that's why the timestamp is not continuous)

```
cat rds2_greece_10C4_karamela.spy.txt | redsea -h --streams |  grep file_contents |  head --lines=1 |  jq --raw-output .rft.file_contents |  base64 --decode > rds2_greece_10C4_karamela.png
```

![Screenshot (481)](https://github.com/user-attachments/assets/c18bc5ab-0a75-425b-99ba-1f4f4d5764cc)
![Screenshot (482)](https://github.com/user-attachments/assets/9fa2aa3f-dbac-4382-9c31-71f8f44da60d)

**Reply by windytan** [reply]
So cool! So it seems to be working. What I would still like to have is a confirmation whether the CRC check failed or succeeded. But this would require us to delay the picture until all CRC parts have been received so we can attach a crc_ok field. So I might save it for a later update to rethink how it would work, if at all.

**Comment by andimik** [comment]
I just had contact with the Greek technician of **Radio Rainbow** (89.0 in Thessaloniki, Greece), who gave me his RDSSpy recording. He also said, it's quite large (17 kB).

![102A_rft_file](https://github.com/user-attachments/assets/bf11a5b8-a7b1-482c-951e-a9bd93a41db1)
[102A - 2002-03-29 09-48-34.zip](https://github.com/user-attachments/files/19328464/102A.-.2002-03-29.09-48-34.zip)

```
cat 102A\ -\ 2002-03-29\ 09-48-34.spy | redsea -h --streams | grep file_contents |  head --lines=1 |  jq --raw-output .rft.file_contents |  base64 --decode > /tmp/102A_rft_file.png
```

**Comment by windytan** [comment]
A few things still to be done before we can call it a release (1.2 probably):
* Needs to be organized a little differently, like it's organized in RDS:

      rft           # Everything RFT
      rft.data      # Things that were received in a data group (file_contents, byte_address ...)
      rft.file_info # File info group (version, id, size, ...)
      rft.crc_info  # CRC info group (CRC mode, chunk CRC, ...)

* We still need to get the error rate down, probably we need to rotate the RDS2 streams according to 4.10 in the specs.
* Would be nice to get some good use for the CRC. It's not used by many stations, but wherever it is, could be a check included in redsea.

**Comment by andimik** [comment]
Just an additional confirmation: long PS works, even with Greek characters. See spy-file above.

```
more /tmp/rainbow.json | grep long | sort | uniq -c
    678 {"pi":"0x102A","bler":0,"group":"15A","tp":true,"prog_type":"Rock music","long_ps":"89 RAINBOW ΘΕΣΣΑΛΟΝΙΚH"}
      2 {"pi":"0x102A","bler":2,"group":"15A","tp":true,"prog_type":"Rock music","long_ps":"89 RAINBOW ΘΕΣΣΑΛΟΝΙΚH"}
```

**Reply by windytan** [reply]
That's great! I found one more problem there. With --show-partial it tries to show partial UTF-8 characters and this prints errors and discards the whole group. Need to only discard the long_ps field in that case.
